File: src-tauri/src/services/quick_note.rs
use crate::database::{get_entity_tags, remove_search_index, set_entity_tags, upsert_search_index, Database};
use crate::error::{AppError, AppResult};
use crate::models::quick_note::{CreateQuickNoteInput, QuickNote};
use chrono::{DateTime, Utc};
use rusqlite::{params, OptionalExtension};
use uuid::Uuid;

pub struct QuickNoteService<'a> {
    db: &'a Database,
}

impl<'a> QuickNoteService<'a> {
    pub fn new(db: &'a Database) -> Self {
        Self { db }
    }

    pub fn create(&self, input: CreateQuickNoteInput) -> AppResult<QuickNote> {
        let id = Uuid::new_v4().to_string();
        let now = Utc::now();
        let tags = input.tags.unwrap_or_default();

        self.db.with_conn(|conn| {
            let tx = conn.unchecked_transaction()?;
            tx.execute(
                "INSERT INTO quick_notes (id, content, note_type, created_at)
                 VALUES (?1, ?2, ?3, ?4)",
                params![id, input.content, input.note_type, now.to_rfc3339()],
            )?;
            set_entity_tags(&tx, "quick_note_tags", "quick_note_id", &id, &tags)?;
            upsert_search_index(&tx, "quick_note", &id, "Quick Note", &input.content)?;
            tx.commit()?;
            Ok(())
        })?;

        self.get(&id)
    }

    pub fn get(&self, id: &str) -> AppResult<QuickNote> {
        self.db.with_conn(|conn| {
            let row = conn
                .query_row(
                    "SELECT id, content, note_type, created_at FROM quick_notes WHERE id = ?1",
                    params![id],
                    |row| {
                        Ok((
                            row.get::<_, String>(0)?,
                            row.get::<_, String>(1)?,
                            row.get::<_, Option<String>>(2)?,
                            row.get::<_, String>(3)?,
                        ))
                    },
                )
                .optional()?
                .ok_or_else(|| AppError::NotFound(format!("quick note {id}")))?;

            let tags = get_entity_tags(conn, "quick_note_tags", "quick_note_id", id)?;
            Ok(QuickNote {
                id: row.0,
                content: row.1,
                note_type: row.2,
                created_at: DateTime::parse_from_rfc3339(&row.3)
                    .map(|d| d.with_timezone(&Utc))
                    .unwrap_or_else(|_| Utc::now()),
                tags,
            })
        })
    }

    pub fn list(&self, limit: Option<i32>) -> AppResult<Vec<QuickNote>> {
        self.db.with_conn(|conn| {
            let sql = match limit {
                Some(n) => format!(
                    "SELECT id, content, note_type, created_at FROM quick_notes
                     ORDER BY created_at DESC LIMIT {n}"
                ),
                None => "SELECT id, content, note_type, created_at FROM quick_notes
                          ORDER BY created_at DESC".to_string(),
            };
            let mut stmt = conn.prepare(&sql)?;
            let rows = stmt.query_map([], |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, String>(1)?,
                    row.get::<_, Option<String>>(2)?,
                    row.get::<_, String>(3)?,
                ))
            })?;
            let mut notes = Vec::new();
            for row in rows {
                let (id, content, note_type, created_at) = row?;
                let tags = get_entity_tags(conn, "quick_note_tags", "quick_note_id", &id)?;
                notes.push(QuickNote {
                    id,
                    content,
                    note_type,
                    created_at: DateTime::parse_from_rfc3339(&created_at)
                        .map(|d| d.with_timezone(&Utc))
                        .unwrap_or_else(|_| Utc::now()),
                    tags,
                });
            }
            Ok(notes)
        })
    }

    pub fn delete(&self, id: &str) -> AppResult<()> {
        self.db.with_conn(|conn| {
            let tx = conn.unchecked_transaction()?;
            tx.execute("DELETE FROM quick_notes WHERE id = ?1", params![id])?;
            remove_search_index(&tx, "quick_note", id)?;
            tx.commit()?;
            Ok(())
        })
    }
}

File: server/internal/core/quicknote.go
package core

import (
	"database/sql"
	"time"
)

// QuickNote mirrors models/quick_note.rs.
type QuickNote struct {
	ID        string    `json:"id"`
	Content   string    `json:"content"`
	NoteType  *string   `json:"noteType,omitempty"`
	CreatedAt time.Time `json:"createdAt"`
	Tags      []string  `json:"tags"`
}

type CreateQuickNoteInput struct {
	Content  string    `json:"content"`
	NoteType *string   `json:"noteType,omitempty"`
	Tags     *[]string `json:"tags,omitempty"`
}

type quickNoteService struct{ db *DB }

func newQuickNoteService(db *DB) *quickNoteService { return &quickNoteService{db: db} }

func (s *quickNoteService) create(in CreateQuickNoteInput) (QuickNote, error) {
	id := newID()
	now := nowUTC()
	tags := []string{}
	if in.Tags != nil {
		tags = *in.Tags
	}
	tx, err := s.db.sql.Begin()
	if err != nil {
		return QuickNote{}, err
	}
	defer tx.Rollback()
	if _, err := tx.Exec(
		`INSERT INTO quick_notes (id, content, note_type, created_at) VALUES (?, ?, ?, ?)`,
		id, in.Content, in.NoteType, rfc3339(now)); err != nil {
		return QuickNote{}, err
	}
	if err := setEntityTags(tx, "quick_note_tags", "quick_note_id", id, tags); err != nil {
		return QuickNote{}, err
	}
	if err := upsertSearchIndex(tx, "quick_note", id, "Quick Note", in.Content); err != nil {
		return QuickNote{}, err
	}
	if err := tx.Commit(); err != nil {
		return QuickNote{}, err
	}
	return s.get(id)
}

func (s *quickNoteService) get(id string) (QuickNote, error) {
	var (
		n        QuickNote
		noteType sql.NullString
		created  string
	)
	err := s.db.sql.QueryRow(
		`SELECT id, content, note_type, created_at FROM quick_notes WHERE id = ?`, id).
		Scan(&n.ID, &n.Content, &noteType, &created)
	if err == sql.ErrNoRows {
		return QuickNote{}, notFound("quick note " + id)
	}
	if err != nil {
		return QuickNote{}, err
	}
	n.NoteType = nullStr(noteType)
	n.CreatedAt = parseRFC3339(created)
	tags, err := getEntityTags(s.db.sql, "quick_note_tags", "quick_note_id", id)
	if err != nil {
		return QuickNote{}, err
	}
	n.Tags = tags
	return n, nil
}

func (s *quickNoteService) list(limit *int) ([]QuickNote, error) {
	q := `SELECT id, content, note_type, created_at FROM quick_notes ORDER BY created_at DESC`
	var args []any
	if limit != nil {
		q += ` LIMIT ?`
		args = append(args, *limit)
	}
	rows, err := s.db.sql.Query(q, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	type staged struct {
		n QuickNote
	}
	var items []QuickNote
	for rows.Next() {
		var (
			n        QuickNote
			noteType sql.NullString
			created  string
		)
		if err := rows.Scan(&n.ID, &n.Content, &noteType, &created); err != nil {
			return nil, err
		}
		n.NoteType = nullStr(noteType)
		n.CreatedAt = parseRFC3339(created)
		items = append(items, n)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	out := []QuickNote{}
	for _, n := range items {
		tags, err := getEntityTags(s.db.sql, "quick_note_tags", "quick_note_id", n.ID)
		if err != nil {
			return nil, err
		}
		n.Tags = tags
		out = append(out, n)
	}
	return out, nil
}

func (s *quickNoteService) delete(id string) error {
	tx, err := s.db.sql.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()
	if _, err := tx.Exec("DELETE FROM quick_notes WHERE id = ?", id); err != nil {
		return err
	}
	if err := removeSearchIndex(tx, "quick_note", id); err != nil {
		return err
	}
	return tx.Commit()
}
